File: backend/formatters.py
"""
Formatting and Summarization Utilities for J.A.D.A Agent.
Handles payload pre-summarization, text truncation, and Markdown fallback formatting.
"""

import json
import re
from typing import Any, Union, Tuple, Optional

MAX_TOOL_OUTPUT_CHARS = 12000


def summarize_paint_defects_data(data: list) -> str:
    """Pre-summarize door-level paint defect inspection arrays into clean Markdown tables."""
    total = len(data)
    fails = [d for d in data if d.get("is_fail") == 1 or str(d.get("door_grade")).lower() == "fail"]
    passes = [d for d in data if str(d.get("door_grade")).lower() == "pass"]

    defects = {}
    for d in fails:
        reason = d.get("reason_description") or "Unspecified Defect"
        defects[reason] = defects.get(reason, 0) + 1

    lines = [
        "## Paint Inspection Defect Summary Data\n",
        f"- **Total Inspected Doors**: {total}",
        f"- **Passes**: {len(passes)}",
        f"- **Failures**: {len(fails)} ({round(len(fails)/total*100, 1) if total else 0}% failure rate)\n",
        "### Defect Breakdown by Type:"
    ]

    if defects:
        for reason, count in sorted(defects.items(), key=lambda x: x[1], reverse=True):
            lines.append(f"- **{reason}**: {count}")
    else:
        lines.append("No defect failures recorded in this dataset.")

    return "\n".join(lines)


def summarize_raw_json_if_needed(raw_str: str) -> str:
    """
    Pre-summarize raw JSON payloads (e.g., paint defects, workorder tracking)
    into structured Markdown summaries before returning to the model.
    """
    if not raw_str or not isinstance(raw_str, str):
        return str(raw_str)

    clean_raw = raw_str.strip()
    if clean_raw in ("null", "[]"):
        return "No records found for the requested parameters."

    if clean_raw.startswith("[") and clean_raw.endswith("]"):
        try:
            data = json.loads(clean_raw)
            if isinstance(data, list) and len(data) > 0 and isinstance(data[0], dict):
                # Paint defects inspection record detection
                if "door_grade" in data[0] or "reason_description" in data[0] or "is_fail" in data[0]:
                    return summarize_paint_defects_data(data)
        except Exception:
            pass

    return raw_str


def truncate_tool_output(
    result: Any,
    max_chars: int = MAX_TOOL_OUTPUT_CHARS,
    response_format: Optional[str] = None
) -> Any:
    """
    Guarantees tool output string is summarized and capped to max_chars.
    If response_format == 'content_and_artifact' or result is a tuple, guarantees returning a 2-tuple (content_str, artifact).
    """
    artifact = None
    if isinstance(result, tuple) and len(result) == 2:
        content, artifact = result
        res_str = content if isinstance(content, str) else json.dumps(content, default=str)
    else:
        res_str = result if isinstance(result, str) else json.dumps(result, default=str)

    res_str = summarize_raw_json_if_needed(res_str)

    if len(res_str) > max_chars:
        res_str = (
            res_str[:max_chars]
            + f"\n\n... (tool output truncated from {len(res_str)} characters to fit context window)"
        )

    if response_format == "content_and_artifact" or isinstance(result, tuple):
        return (res_str, artifact)

    return res_str


def format_fallback_tool_summary(raw_output: str) -> str:
    """
    Format raw JSON or search tool output into a clean Markdown summary
    if the LLM failed to generate tokens.
    """
    if not raw_output or not isinstance(raw_output, str):
        return "Task completed successfully."

    clean_raw = raw_output.strip()

    # 1. Try paint defects
    try:
        data = json.loads(clean_raw)
        if isinstance(data, list) and len(data) > 0 and isinstance(data[0], dict):
            if "is_fail" in data[0] or "door_number" in data[0]:
                return summarize_paint_defects_data(data)
    except Exception:
        pass

    # 2. Weather snippet formatting
    if any(k in clean_raw for k in ["temp_f", "temp_c", "humidity"]):
        temp_f_match = re.search(r"\"temp_f\"\s*:\s*([\d\.]+)", clean_raw)
        humidity_match = re.search(r"\"humidity\"\s*:\s*(\d+)", clean_raw)
        condition_match = re.search(r"\"text\"\s*:\s*\"([^\"]+)\"", clean_raw)

        lines = ["## Weather Summary\n"]
        if temp_f_match:
            val_f = float(temp_f_match.group(1))
            val_c = round((val_f - 32) * 5 / 9, 1)
            lines.append(f"- **Temperature**: {val_f}°F ({val_c}°C)")
        if humidity_match:
            lines.append(f"- **Humidity**: {humidity_match.group(1)}%")
        if condition_match:
            lines.append(f"- **Conditions**: {condition_match.group(1).strip()}")

        if len(lines) > 1:
            return "\n".join(lines)

    # 3. Clean raw fallback
    cleaned = re.sub(r"[\r\n]{3,}", "\n\n", clean_raw).strip()
    return cleaned or "The tool query completed successfully."

File: backend_go/pkg/formatters/formatters.go
package formatters

import (
	"encoding/json"
	"fmt"
	"math"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

const MaxToolOutputChars = 12000

func SummarizePaintDefectsData(data []map[string]interface{}) string {
	total := len(data)
	var fails []map[string]interface{}
	var passes []map[string]interface{}

	defects := make(map[string]int)

	for _, d := range data {
		isFail := false
		if val, ok := d["is_fail"]; ok {
			switch v := val.(type) {
			case float64:
				if v == 1 {
					isFail = true
				}
			case int:
				if v == 1 {
					isFail = true
				}
			case bool:
				if v {
					isFail = true
				}
			}
		}

		if grade, ok := d["door_grade"]; ok {
			if strings.EqualFold(fmt.Sprintf("%v", grade), "fail") {
				isFail = true
			} else if strings.EqualFold(fmt.Sprintf("%v", grade), "pass") {
				passes = append(passes, d)
			}
		}

		if isFail {
			fails = append(fails, d)
			reason := "Unspecified Defect"
			if r, ok := d["reason_description"]; ok && fmt.Sprintf("%v", r) != "" && fmt.Sprintf("%v", r) != "<nil>" {
				reason = fmt.Sprintf("%v", r)
			}
			defects[reason]++
		}
	}

	failRate := 0.0
	if total > 0 {
		failRate = math.Round(float64(len(fails))/float64(total)*1000) / 10.0
	}

	var lines []string
	lines = append(lines, "## Paint Inspection Defect Summary Data\n")
	lines = append(lines, fmt.Sprintf("- **Total Inspected Doors**: %d", total))
	lines = append(lines, fmt.Sprintf("- **Passes**: %d", len(passes)))
	lines = append(lines, fmt.Sprintf("- **Failures**: %d (%.1f%% failure rate)\n", len(fails), failRate))
	lines = append(lines, "### Defect Breakdown by Type:")

	if len(defects) > 0 {
		type kv struct {
			Key   string
			Value int
		}
		var sortedDefects []kv
		for k, v := range defects {
			sortedDefects = append(sortedDefects, kv{k, v})
		}
		sort.Slice(sortedDefects, func(i, j int) bool {
			return sortedDefects[i].Value > sortedDefects[j].Value
		})

		for _, item := range sortedDefects {
			lines = append(lines, fmt.Sprintf("- **%s**: %d", item.Key, item.Value))
		}
	} else {
		lines = append(lines, "No defect failures recorded in this dataset.")
	}

	return strings.Join(lines, "\n")
}

func SummarizeRawJSONIfNeeded(rawStr string) string {
	if strings.TrimSpace(rawStr) == "" {
		return rawStr
	}

	cleanRaw := strings.TrimSpace(rawStr)
	if cleanRaw == "null" || cleanRaw == "[]" {
		return "No records found for the requested parameters."
	}

	if strings.HasPrefix(cleanRaw, "[") && strings.HasSuffix(cleanRaw, "]") {
		var list []map[string]interface{}
		if err := json.Unmarshal([]byte(cleanRaw), &list); err == nil && len(list) > 0 {
			first := list[0]
			_, hasGrade := first["door_grade"]
			_, hasReason := first["reason_description"]
			_, hasIsFail := first["is_fail"]
			if hasGrade || hasReason || hasIsFail {
				return SummarizePaintDefectsData(list)
			}
		}
	}

	return rawStr
}

func TruncateToolOutput(result interface{}, maxChars int) string {
	if maxChars <= 0 {
		maxChars = MaxToolOutputChars
	}

	var resStr string
	switch v := result.(type) {
	case string:
		resStr = v
	default:
		bytes, err := json.Marshal(v)
		if err != nil {
			resStr = fmt.Sprintf("%v", v)
		} else {
			resStr = string(bytes)
		}
	}

	resStr = SummarizeRawJSONIfNeeded(resStr)

	if len(resStr) > maxChars {
		originalLen := len(resStr)
		resStr = resStr[:maxChars] + fmt.Sprintf("\n\n... (tool output truncated from %d characters to fit context window)", originalLen)
	}

	return resStr
}

func FormatFallbackToolSummary(rawOutput string) string {
	if strings.TrimSpace(rawOutput) == "" {
		return "Task completed successfully."
	}

	cleanRaw := strings.TrimSpace(rawOutput)

	// 1. Check paint defects JSON
	if strings.HasPrefix(cleanRaw, "[") && strings.HasSuffix(cleanRaw, "]") {
		var list []map[string]interface{}
		if err := json.Unmarshal([]byte(cleanRaw), &list); err == nil && len(list) > 0 {
			first := list[0]
			_, hasIsFail := first["is_fail"]
			_, hasDoorNum := first["door_number"]
			if hasIsFail || hasDoorNum {
				return SummarizePaintDefectsData(list)
			}
		}
	}

	// 2. Weather snippet formatting
	if strings.Contains(cleanRaw, "temp_f") || strings.Contains(cleanRaw, "temp_c") || strings.Contains(cleanRaw, "humidity") {
		reTempF := regexp.MustCompile(`"temp_f"\s*:\s*([\d\.]+)`)
		reHumidity := regexp.MustCompile(`"humidity"\s*:\s*(\d+)`)
		reCondition := regexp.MustCompile(`"text"\s*:\s*"([^"]+)"`)

		matchF := reTempF.FindStringSubmatch(cleanRaw)
		matchHum := reHumidity.FindStringSubmatch(cleanRaw)
		matchCond := reCondition.FindStringSubmatch(cleanRaw)

		var lines []string
		lines = append(lines, "## Weather Summary\n")

		if len(matchF) > 1 {
			valF, _ := strconv.ParseFloat(matchF[1], 64)
			valC := math.Round((valF-32)*5/9*10) / 10
			lines = append(lines, fmt.Sprintf("- **Temperature**: %.1f°F (%.1f°C)", valF, valC))
		}
		if len(matchHum) > 1 {
			lines = append(lines, fmt.Sprintf("- **Humidity**: %s%%", matchHum[1]))
		}
		if len(matchCond) > 1 {
			lines = append(lines, fmt.Sprintf("- **Conditions**: %s", strings.TrimSpace(matchCond[1])))
		}

		if len(lines) > 1 {
			return strings.Join(lines, "\n")
		}
	}

	reNewlines := regexp.MustCompile(`[\r\n]{3,}`)
	cleaned := reNewlines.ReplaceAllString(cleanRaw, "\n\n")
	cleaned = strings.TrimSpace(cleaned)

	if cleaned != "" {
		return cleaned
	}
	return "The tool query completed successfully."
}
